pkg/gadgets/ebpftop/gadget.go

```diff
@@ -0,0 +1,255 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpftop
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/cilium/ebpf"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+	"github.com/kinvolk/inspektor-gadget/pkg/bpfstats"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/ebpftop/piditer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/ebpftop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+
+	started  bool
+	donechan chan bool
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+}
+
+func NewFactory() gadgets.TraceFactory {
+	return &TraceFactory{
+		BaseFactory: gadgets.BaseFactory{DeleteTrace: deleteTrace},
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	t := `ebpftop shows cpu time used by ebpf programs.
+
+The following parameters are supported:
+ - %s: Output interval, in seconds. (default %d)
+ - %s: Maximum rows to print. (default %d)
+ - %s: The field to sort the results by (%s). (default %s)`
+	return fmt.Sprintf(t, types.IntervalParam, types.IntervalDefault,
+		types.MaxRowsParam, types.MaxRowsDefault,
+		types.SortByParam, strings.Join(types.SortBySlice, ","), types.SortByDefault)
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	// TODO: need to stop something in here?
```

Comment:
The trace could be deleted without `Stop()` being called before, so if the trace was started, you need to stop the Go routine (`t.donechan <- true`) and stop the stats (`bpfstats.DisableBPFStats`) like you wrote in `Stop()`.

pkg/gadgets/network-graph/gadget.go gives an example of code where some code in `Stop()` has been factorized in `stop()` to avoid code duplication in `deleteTrace()`. Some other gadgets duplicated the code in `Stop()` and `deleteTrace()`.

Reply:
Thanks! This should be handled correctly, now.